### ISSUE TITLE
Dynamic surfaces

### ISSUE THREAD

**Issue by dijana-sagit** [facebook/react-360#482]
Hi, my question is related to the surface rendering (particularly flat). According to the example _MultiRoot_, you render all the surfaces from the _client.js_. Once rendered, is there a way to update the location of the surface from the code? Or, is there a way to create and render a surface on the go from the code? I'm looking into implementing a 2D popup when I click on a _VrButton_.
Thanks!

**Comment by andrewimm** [comment]
Yes, the same setAngle methods you use to initially position your flat surface can be called anytime to reposition them. The same is true for resizing or re-shaping a surface. This needs to be handled from the client.js side, but you can easily implement your behavior with a Native Module that either creates a surface on demand, or (recommended) previously creates a surface with no contents, and repositions it when you need it.

As a team, we are working on developing an official API for repositioning or generating surfaces from React, but we want to make sure that it is flexible enough to support all potential use cases we have in mind – we don't want to change APIs and break code later. In the meantime, we recommend the path of creating your own native module methods.

**Comment by flyandi** [comment]
how do u access the actual Instance from a NativeModule? All I am getting is the context that doesn't seem to expose access to the actual r360 instance.

**Comment by andrewimm** [comment]
You can't send the value through the native module – you can only transfer primitives and objects that follow the structured clone algorithm. What you can do is expose a method to React that updates a surface instance on the client side. Building a specific module for your application is easier; building a generalized API that supports every possible case today, and doesn't prevent future behavior tomorrow, is going to be a little trickier.

As we look forward at possible APIs, we are considering the ability to register each surface with a unique identifier. This will be used to reference it from the React side – a common pattern in React Native setups like this. From there, you could trigger a variety of reshaping and repositioning effects, as long as they were exposed through the module.

```
const {SurfaceModule} = NativeModules;
SurfaceModule.reposition('mysurface', {angleHoriz: -0.5, angleVert: 0.3});
```

**Comment by flyandi** [comment]
So if I am understanding this right, all surfaces are created in client.js only during initialization  and although we can modify a surfaces behavior, there is no way of creating dynamic surfaces at this point since that would require re-running client.js.. correct?

**Comment by andrewimm** [comment]
No, surfaces can be created anytime. If you look at the the code in the MultiRoot sample, those surfaces are actually being created after React has initialized. At any point in app initialization, a surface can be generated and used with `renderToSurface()` to create a new React root.

While it's possible to generate Surfaces on the fly, I do recommend you pre-allocate and reuse them in situations where you know the number of surfaces ahead of time. For instance, to show a display every time a button is clicked, you could send a message via Native Module to generate a new popup each time – but that's wasteful of resources. It'd be better to preallocate a popup surface, and then reposition and reshape it when needed. Surfaces are cheap to create today, but we have some optimizations we're looking at in the near future that would dramatically increase the visual quality of surface content, similar to how the Gear VR re-release in early 2017 effectively doubled the pixel density with only a software update. These optimizations, while great for app quality, would make it more expensive to generate surfaces.

I also want to highlight a new Sample I just published that shows surfaces being smoothly repositioned each frame: https://github.com/facebook/react-360/tree/master/Samples/HeadlockedSurfaces

**Comment by andrewimm** [comment]
I haven't stress-tested, but I don't think there's a significant variable cost to adding more surfaces. If you have 5, I wouldn't worry about perf. I've done some hotspot demos that had 12 running on the screen with no measurable issues.

If you're going to end up with a lot, you may want to build a Native Module that implements a resource pool for Surfaces. I can describe below vaguely how I'd implement it, but it'd probably be helpful to get a demo of this in the Samples dir.

It comes down to three pieces: a method to request a surface, a method that tells client.js to render a specific root to a surface, and a method to reposition a surface.

From React, those might look vaguely like this:
```
const {SurfaceModule} = NativeModules;

SurfaceModule.requestSurface().then(handle => {
  // handle is a unique string identifier for referencing a surface
  // Resize to 500x300px
  SurfaceModule.resize(handle, 500, 300);
  // Tell client.js to instantiate a new instance of <ReactRootName> on the new surface
  SurfaceModule.renderToSurface('ReactRootName', { /* initial props */ }, handle);

  // Later on, free up the surface again
  SurfaceModule.unmountComponentAtSurface(handle);
});
```

On the client.js side, you'd implement a pool of declared-but-unused surfaces. When a surface is requested, you either pull one out of the pool, or create a new one. There's still some kinks in the model I've outlined above, like whether a surface gets released when you unmount its contents, or whether it needs to be explicitly released – that's why these APIs take some time to evolve. Also on that note, I don't think we've actually exposed the functionality to unmount surface content from client.js, but I'll get on that today.

I'm looking to build at least some of these as experimental APIs in the next week. My team has had a lot of discussions about this space, since it's also relevant to our internal Oculus use cases.

**Comment by andrewimm** [comment]
I think there might still be a disconnect here. A native module is simply a way to trigger something in client.js from within React, they do not automatically contain anything. The method names are arbitrary, I have just chosen to copy the API because it creates a sense of familiarity.

More on native modules, including a sample:
https://facebook.github.io/react-360/docs/native-modules.html

**Comment by ryanf12** [comment]
Thanks for posting the Headlocked sample, seeing working samples is helpful.  

I'm still not seeing exactly _how_ the the updates are rendered.  

I tried incorporating this snippet to a surface resize, and it doesn't change the size:  

frame: () => {...}

Here is another example where the values correctly get pulled from an array, made into buttons with .map, and placed into the Surface.  You can see the variables correctly get pushed into Surface, you can verify this by looking at the console output for frontPanel._width and frontPanel._height.  

The updates to the sizes on the surface correctly get sent from the array to the surface, the updates just aren't rendering.    

**Client.js**

    import {ReactInstance, Surface} from 'react-360-web';
    import panelSizes from './panelSizes';
    
    
    export const frontPanel = new Surface(4096, 350, Surface.SurfaceShape.Cylinder);
    
    function init(bundle, parent, options = {}) {
    
        frontPanel.setAngle(0,  0); //Puts the cylinder in front 
          frontPanel.resize(1024, 512); //Resizes the cylinder 
        
      const r360 = new ReactInstance(bundle, parent, {
        frame: () => {
          frontPanel.resize(50, 50); // <-- The headlocked example uses something similar to this, but this here doesn't actually do anything.  
        },
        fullScreen: true,
        ...options,
      });
    
    
        r360.renderToSurface(
            r360.createRoot('resize'),
            frontPanel,
        );
    
      r360.compositor.setBackground(r360.getAssetURL('360_world.jpg'));
    }
    
    window.React360 = {init};
     

index.js 

    import React from 'react';
    import {
      AppRegistry,
      StyleSheet,
      Text,
      View,
      VrButton, 
    } from 'react-360';
    
    import panelSizes from './panelSizes';
    import {frontPanel} from './client'; 
    
    export default class resize extends React.Component {
        constructor(props) {
          super(props);
          this.state = {currentPage: 0}
        }
        
        render() {
            
        const currentPage = this.state.currentPage; 
    
        changePage = (e) => {
            this.setState({ currentPage : e});
            console.log("Before Resize width: " + frontPanel._width + " height: " + frontPanel._height)
            frontPanel.resize(panelSizes[e].width, panelSizes[e].height); // <-- This correctly passes the size values from the array into the surface, but the page doesn't render the updates.   
            console.log("After Resize width: " + frontPanel._width + " height: " + frontPanel._height) 
    //Right here something needs to force a re-render of the surface.   
        }
    
        const buttons = panelSizes.map((button, index) =>
            <View key={index} style={{position:"absolute", top:panelSizes[index].top, left:panelSizes[index].left}}>
                <VrButton onClick={changePage.bind(this, panelSizes[index].targetState)}>
                    <Text style={{padding:20, backgroundColor:'rgba(255, 255, 255, .2)', textAlignVertical: "center",textAlign: "center"}}>Go to state {index}</Text>
                </VrButton>
            </View>
        );
          
        return (
            <View style={styles.panel}>
            <Text>The current state is:  {this.state.currentPage}</Text>
            <Text>The size of the surface is:  W: {frontPanel._width} x H: {frontPanel._height}</Text>
                {buttons}
            </View>
        );
      }
    };
    
    const styles = StyleSheet.create({
        panel: {
            // Fill the entire surface
            width: 1000,
            height: 600,
            backgroundColor: 'rgba(255, 255, 255, 0.4)',
            justifyContent: 'center',
            alignItems: 'center',
        }
    });
    
    AppRegistry.registerComponent('resize', () => resize);
     
panelsizes.js 

    var panelSizes = [{
                "title": "Button 1", 
                "top":"30", 
                "left":"33", 
                "height": "512", 
                "width": "1024", 
                "targetState":"0", 
            },{ 
                "title": "Button 1", 
                "top":"30", 
                "left":"52", 
                "height": "128", 
                "width": "256", 
                "targetState":"1", 
            }]; 
                
    export default panelSizes;  


The resizing of the surface will only happen with an event such as a click, is there any way to explicitly tell React 360 to update a surface on an onClick event?  Everything about this seems like it should work.  Anything such as a "surface.forceUpdate()" function?

**Comment by devsatish** [comment]
@andrewimm Hello Andrew; this is regarding referencing Surface in index.js. In my example; I am creating surfaces and a native module - "SurfaceModule". All I am trying to do is a reposition a surface , by getting a reference to the surface by the name; it doesnt seem to work; What's the api to reference a surface by name ?

```
//client.js
import {ReactInstance, Surface} from 'react-360-web';
import {Module} from 'react-360-web';
function init(bundle, parent, options = {}) {
  const r360 = new ReactInstance(bundle, parent, {
    fullScreen: true,
    nativeModules: [
      new SurfaceModule(),
    ],
    ...options,
  });

  const s1 = new Surface(256, 256, Surface.SurfaceShape.Flat);
  s1.setAngle(0,1.5707963267949); //pi/4
  r360.renderToSurface(
    r360.createRoot('MultiPanel', { lbl:'Cam1',}),
    s1
  );
  r360.compositor.setBackground(r360.getAssetURL('360_world.jpg'));
}

class SurfaceModule extends Module {
  constructor() {
        console.log('constructor - SurfaceModule');
		super('surfaceModule');
  }
  triggerMe(surfaceName, horizontalAngle, verticalAngle) {
    s1.setAngle(horizontalAngle,verticalAngle);
  }
}
window.React360 = {init};


//index.js
import React from 'react';
import {AppRegistry,
  StyleSheet,
  Text,
  View,
  VrButton,
  Image,
  asset,
  uri,
  Environment,
} from 'react-360';
import {
  NativeModules,
} from 'react-360';
const surfaceModule = NativeModules.SurfaceModule;

export default class MultiPanel extends React.Component {

 constructor(props) {
    super();
  }

  doClick = () => {
   console.log(surfaceModule);
    surfaceModule.trigger('s1', 1.5708, 0.7854);
  }

  
  render() {
    return (
         <View style={{flexDirection: 'column', height: 100, padding: 20}}>
         <Text style={{textAlign: 'center', color:'pink'}}>{this.props.lbl}</Text>
          <VrButton onClick={this.doClick}>
              <Image style={{
                  height: 128,
                  width: 128,
                }}  source={{uri: './static_assets/hotspot.png'}}/>
        </VrButton>
      </View>

      
    );
  }
};
AppRegistry.registerComponent('MultiPanel', () => MultiPanel);
```

**Comment by balintlaczko** [comment]
Dear all, I would like to do this:
when clicked on a VrButton in index.js -> setangle of the Surface on which the VrButton is rendered in client.js.
I understand that I probably have to set up an event listener in the client and fire it in the index (right...?), but I am a beginner on this and got stuck with this - which feels to be an easy thing to achieve I guess if you know what you're doing, so it would be so nice if some of you could throw me an example how to do this! Would be eternally grateful!!
:)
